transformations/transformations/block_index_transformations.py

```diff
@@ -0,0 +1,428 @@
+    dummy-argument, the block-index is appended to the argument's dimensions. It should be noted that this logic relies
+    on the :any:`CallStatement` being free of any sequence-association.
+
+    For example, the following code:
+
+    .. code-block:: fortran
+
+        subroutine kernel1(nblks, ...)
+           ...
+           integer, intent(in) :: nblks
+           integer :: ibl
+           real :: var(jlon,nlev,nblks)
+
+           do ibl=1,nblks
+             do jlon=1,nproma
+               var(jlon,:) = 0.
+             enddo
+
+             call kernel2(var,...)
+           enddo
+           ...
+        end subroutine kernel1
+
+        subroutine kernel2(var, ...)
+           ...
+           real :: var(jlon,nlev)
+        end subroutine kernel2
+
+    is transformed to:
+
+    .. code-block:: fortran
+
+        subroutine kernel1(nblks, ...)
+           ...
+           integer, intent(in) :: nblks
+           integer :: ibl
+           real :: var(jlon,nlev,nblks)
+
+           do ibl=1,nblks
+             do jlon=1,nproma
+               var(jlon,:,ibl) = 0.
+             enddo
+
+             call kernel2(var(:,:,ibl),...)
+           enddo
+           ...
+        end subroutine kernel1
+
+        subroutine kernel2(var, ...)
+           ...
+           real :: var(jlon,nlev)
+        end subroutine kernel2
+
+    Specific arrays in individual routines can also be marked for exclusion from this transformation by assigning
+    them to the `exclude_arrays` list in the :any:`SchedulerConfig`.
+
+    Parameters
+    ----------
+    block_dim : :any:`Dimension`
+        :any:`Dimension` object describing the variable conventions used in code
+        to define the blocking data dimension and iteration space.
+    key : str, optional
+        Specify a different identifier under which trafo_data is stored
+    """
+
+    _key = 'BlockIndexInjectTransformation'
+    """Default identifier for trafo_data entry"""
+
+    # This trafo only operates on procedures
+    item_filter = (ProcedureItem,)
+
+    def __init__(self, block_dim, key=None):
+        self.block_dim = block_dim
+        if key:
+            self._key = key
+
+    def transform_subroutine(self, routine, **kwargs):
+
+        role = kwargs['role']
+        targets = tuple(str(t).lower() for t in as_tuple(kwargs.get('targets', None)))
+
+        exclude_arrays = []
+        if (item := kwargs.get('item', None)):
+            exclude_arrays = item.config.get('exclude_arrays', [])
+
+        if role == 'kernel':
+            self.process_kernel(routine, targets, exclude_arrays)
+
+        #TODO: we also need to process any code inside a loki/acdc parallel pragma at the driver layer
+
+    @staticmethod
+    def _update_expr_map(var, rank, index):
+        """
+        Return a map with the block-index appended to the variable's dimensions.
+        """
+
+        if getattr(var, 'dimensions', None):
```

Comment:
Hmm clever way of checking whether `var` is an `Array` and whether `dimensions` is empty `()` ... also a bit confusing 😄

Reply:
As agreed offline, this could probably do with a comment 😅